File: mobile/android/app/src/main/java/com/sunpanel/app/Servers.kt
package com.sunpanel.app

import java.net.URI
import java.net.URISyntaxException
import org.json.JSONArray
import org.json.JSONObject

/** Cách ứng dụng nối tới một máy chủ. */
enum class Kind {
    /** Mở giao diện panel đã cài trên máy chủ. */
    PANEL,
    /** Nối thẳng vào máy chủ qua SSH, không cần cài gì trên đó. */
    SSH;

    companion object {
        /** Đọc kiểu từ dữ liệu đã lưu; bản ghi cũ không có trường này. */
        fun of(text: String?): Kind = if (text.equals("ssh", ignoreCase = true)) SSH else PANEL
    }
}

/** Một máy chủ đã lưu trong ứng dụng. */
data class Server(
    val id: String,
    val name: String = "",
    val kind: Kind = Kind.PANEL,

    /** Địa chỉ panel kèm đường dẫn bí mật. Chỉ dùng cho [Kind.PANEL]. */
    val url: String = "",

    /** Các trường dưới đây chỉ dùng cho [Kind.SSH]. */
    val host: String = "",
    val port: Int = 22,
    val user: String = "",
    /** Nội dung khóa riêng dạng PEM, người dùng dán vào. */
    val privateKey: String = "",
    /** Mật khẩu, chỉ được ghi khi người dùng tự chọn nhớ. */
    val password: String = "",
    /** Vân tay khóa máy chủ SSH đã ghi nhận ở lần kết nối đầu. */
    val hostKey: String = "",

    val last: Boolean = false,
    /**
     * Vân tay SHA-256 của chứng chỉ TLS mà người dùng đã đồng ý tin, dạng hoa
     * không dấu phân cách. Chỉ dùng cho [Kind.PANEL].
     */
    val certFingerprint: String = "",
) {
    /** Dòng địa chỉ hiện dưới tên máy chủ. */
    fun label(): String = if (kind == Kind.SSH) "$user@$host:$port" else url
}

/** Lỗi thông tin máy chủ không dùng được, kèm mã để giao diện tự dịch. */
class InvalidAddress(val reason: Reason) : Exception(reason.name) {
    enum class Reason { EMPTY, MALFORMED, SCHEME, HOST, USER, PORT }
}

/**
 * Toàn bộ phần xử lý danh sách máy chủ, viết thuần Kotlin.
 *
 * Không đụng gì tới Android nên chạy thẳng được trong bài kiểm thử trên JVM —
 * đây là chỗ dễ sai nhất của ứng dụng (địa chỉ gõ tay, dữ liệu cũ trên máy) nên
 * phải kiểm được mà không cần máy ảo.
 */
object Servers {

    /** Kiểm và chuẩn hóa địa chỉ panel người dùng gõ vào. */
    fun normalizeUrl(raw: String): String {
        var text = raw.trim()
        if (text.isEmpty()) throw InvalidAddress(InvalidAddress.Reason.EMPTY)

        // Thiếu giao thức thì hiểu là http: panel chạy HTTP cho tới khi người dùng
        // gắn chứng chỉ, và bắt gõ đủ "http://" chỉ tạo thêm chỗ sai.
        if (!text.contains("://")) text = "http://$text"

        val parsed =
            try {
                URI(text)
            } catch (e: URISyntaxException) {
                throw InvalidAddress(InvalidAddress.Reason.MALFORMED)
            }

        val scheme = parsed.scheme?.lowercase()
        if (scheme != "http" && scheme != "https") throw InvalidAddress(InvalidAddress.Reason.SCHEME)

        val authority = parsed.authority
        if (authority.isNullOrEmpty()) throw InvalidAddress(InvalidAddress.Reason.HOST)

        // Đường dẫn bí mật phải kết thúc bằng dấu gạch chéo, nếu không thẻ base của
        // panel phân giải sai và giao diện nạp tài nguyên từ thư mục cha.
        var path = parsed.rawPath ?: ""
        if (!path.endsWith("/")) path += "/"

        return "$scheme://$authority$path"
    }

    /** Tách máy chủ trong địa chỉ đang mở, để biết trang nào thuộc panel nào. */
    fun sameOrigin(a: String, b: String): Boolean {
        return try {
            val left = URI(a)
            val right = URI(b)
            left.scheme.equals(right.scheme, ignoreCase = true) &&
                left.authority.equals(right.authority, ignoreCase = true)
        } catch (e: URISyntaxException) {
            false
        }
    }

    /** Sinh định danh chưa dùng trong danh sách. */
    fun nextId(list: List<Server>): String {
        var next = 1
        while (list.any { it.id == "s$next" }) next++
        return "s$next"
    }

    /** Thêm mới hoặc cập nhật một máy chủ, trả về danh sách mới. */
    fun save(list: List<Server>, draft: Server): List<Server> {
        val clean = normalize(draft)

        if (clean.id.isEmpty()) {
            return list + clean.copy(id = nextId(list))
        }
        return list.map { server ->
            if (server.id != clean.id) {
                server
            } else {
                // Đổi địa chỉ là đổi máy: vân tay cũ không còn nói lên điều gì về
                // máy mới, giữ lại là tự bịt mắt mình.
                val sameMachine =
                    if (clean.kind == Kind.SSH) server.host == clean.host && server.port == clean.port
                    else sameOrigin(server.url, clean.url)
                clean.copy(
                    last = server.last,
                    certFingerprint = if (sameMachine) server.certFingerprint else "",
                    hostKey = if (sameMachine) server.hostKey else "",
                )
            }
        }
    }

    /** Kiểm và dọn một bản ghi trước khi lưu. */
    fun normalize(draft: Server): Server {
        if (draft.kind == Kind.SSH) {
            val host = draft.host.trim()
            val user = draft.user.trim()
            if (host.isEmpty()) throw InvalidAddress(InvalidAddress.Reason.HOST)
            if (host.any { it.isWhitespace() } || host.contains('/')) {
                throw InvalidAddress(InvalidAddress.Reason.MALFORMED)
            }
            if (user.isEmpty()) throw InvalidAddress(InvalidAddress.Reason.USER)

            val port = if (draft.port == 0) 22 else draft.port
            if (port !in 1..65535) throw InvalidAddress(InvalidAddress.Reason.PORT)

            return draft.copy(
                name = draft.name.trim().ifEmpty { "$user@$host" },
                host = host,
                user = user,
                port = port,
                url = "",
                certFingerprint = "",
            )
        }

        val address = normalizeUrl(draft.url)
        return draft.copy(
            kind = Kind.PANEL,
            name = draft.name.trim().ifEmpty { address },
            url = address,
            host = "",
            user = "",
            privateKey = "",
            password = "",
            hostKey = "",
        )
    }

    /** Ghi nhớ khóa máy chủ SSH sau lần kết nối đầu tiên. */
    fun trustHostKey(list: List<Server>, id: String, fingerprint: String): List<Server> =
        list.map { if (it.id == id) it.copy(hostKey = fingerprint) else it }

    /** Xóa một máy chủ khỏi danh sách. */
    fun remove(list: List<Server>, id: String): List<Server> = list.filter { it.id != id }

    /** Đánh dấu máy chủ vừa mở để lần chạy sau tự vào thẳng. */
    fun markLast(list: List<Server>, id: String): List<Server> = list.map { it.copy(last = it.id == id) }

    /** Ghi nhớ chứng chỉ mà người dùng đã đồng ý tin cho một máy chủ. */
    fun trustCertificate(list: List<Server>, id: String, fingerprint: String): List<Server> =
        list.map { if (it.id == id) it.copy(certFingerprint = fingerprint) else it }

    /** Máy chủ mở gần nhất. */
    fun last(list: List<Server>): Server? = list.firstOrNull { it.last }

    /** Chuyển danh sách thành JSON để lưu. */
    fun encode(list: List<Server>): String {
        val array = JSONArray()
        for (server in list) {
            val item = JSONObject()
            item.put("id", server.id)
            item.put("name", server.name)
            item.put("url", server.url)
            item.put("kind", if (server.kind == Kind.SSH) "ssh" else "panel")
            item.put("host", server.host)
            item.put("port", server.port)
            item.put("user", server.user)
            item.put("key", server.privateKey)
            item.put("password", server.password)
            item.put("hostKey", server.hostKey)
            item.put("last", server.last)
            item.put("cert", server.certFingerprint)
            array.put(item)
        }
        return array.toString()
    }

    /**
     * Đọc danh sách từ JSON đã lưu.
     *
     * Dữ liệu hỏng không được làm ứng dụng không mở lên được: bắt đầu lại với danh
     * sách rỗng còn hơn treo ngay màn hình đầu tiên. Từng mục hỏng cũng vậy — mất
     * một máy chủ nhẹ hơn mất cả danh sách.
     */
    fun decode(text: String?): List<Server> {
        if (text.isNullOrBlank()) return emptyList()

        val array =
            try {
                JSONArray(text)
            } catch (e: Exception) {
                return emptyList()
            }

        val list = mutableListOf<Server>()
        for (i in 0 until array.length()) {
            val item = array.optJSONObject(i) ?: continue
            val id = item.optString("id")
            if (id.isEmpty()) continue

            // Bản ghi lưu từ phiên bản trước chỉ có panel nên không ghi trường kind.
            val kind = Kind.of(item.optString("kind"))
            val url = item.optString("url")
            val host = item.optString("host")
            if (kind == Kind.PANEL && url.isEmpty()) continue
            if (kind == Kind.SSH && host.isEmpty()) continue

            list.add(
                Server(
                    id = id,
                    kind = kind,
                    name = item.optString("name").ifEmpty { if (kind == Kind.SSH) host else url },
                    url = url,
                    host = host,
                    port = item.optInt("port", 22).let { if (it <= 0) 22 else it },
                    user = item.optString("user"),
                    privateKey = item.optString("key"),
                    password = item.optString("password"),
                    hostKey = item.optString("hostKey"),
                    last = item.optBoolean("last", false),
                    certFingerprint = item.optString("cert"),
                )
            )
        }
        return list
    }
}

File: desktop/servers.go
package main

import (
	"encoding/json"
	"errors"
	"net/url"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
)

// Kind là cách ứng dụng nối tới một máy chủ.
type Kind string

const (
	// KindPanel mở giao diện panel đã cài trên máy chủ.
	KindPanel Kind = "panel"
	// KindSSH nối thẳng vào máy chủ qua SSH, không cần cài gì trên đó.
	KindSSH Kind = "ssh"
)

// Server là một máy chủ đã lưu trong ứng dụng.
type Server struct {
	// ID để giao diện tham chiếu tới đúng máy chủ khi sửa hoặc xóa.
	ID string `json:"id"`
	// Name là tên người dùng tự đặt, ví dụ "VPS Sài Gòn".
	Name string `json:"name"`
	// Kind quyết định phần còn lại của bản ghi được đọc theo cách nào.
	Kind Kind `json:"kind"`

	// URL là địa chỉ đầy đủ kèm đường dẫn bí mật của panel. Chỉ dùng cho KindPanel.
	URL string `json:"url,omitempty"`

	// Các trường dưới đây chỉ dùng cho KindSSH.
	Host string `json:"host,omitempty"`
	Port int    `json:"port,omitempty"`
	User string `json:"user,omitempty"`
	// KeyPath trỏ tới tệp khóa riêng trên máy này; để trống là đăng nhập bằng
	// mật khẩu.
	KeyPath string `json:"keyPath,omitempty"`
	// Password chỉ được ghi khi người dùng tự chọn nhớ. Mặc định là không: một
	// tệp trên đĩa không phải chỗ cho mật khẩu root của máy chủ.
	Password string `json:"password,omitempty"`
	// Fingerprint là vân tay khóa máy chủ đã ghi nhận ở lần kết nối đầu.
	Fingerprint string `json:"fingerprint,omitempty"`

	// Last đánh dấu máy chủ mở gần nhất, để lần sau tự mở lại.
	Last bool `json:"last,omitempty"`
}

// Label là dòng địa chỉ hiện dưới tên máy chủ.
func (s Server) Label() string {
	if s.Kind == KindSSH {
		return s.User + "@" + s.Host + ":" + strconv.Itoa(s.sshPort())
	}
	return s.URL
}

// sshPort trả về cổng SSH, mặc định 22.
func (s Server) sshPort() int {
	if s.Port <= 0 {
		return 22
	}
	return s.Port
}

// Store giữ danh sách máy chủ trên đĩa.
//
// Tệp cấu hình nằm trong thư mục cấu hình của người dùng chứ không cạnh binary:
// ứng dụng có thể được chép vào /usr/local/bin hay Program Files, những nơi mà
// một tiến trình người dùng thường không ghi được.
type Store struct {
	path string
	mu   sync.Mutex
	list []Server
}

// NewStore mở kho máy chủ, tạo thư mục nếu chưa có.
func NewStore() (*Store, error) {
	dir, err := os.UserConfigDir()
	if err != nil {
		return nil, err
	}
	dir = filepath.Join(dir, "sunpanel")
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return nil, err
	}
	return openStore(filepath.Join(dir, "servers.json"))
}

// openStore mở kho tại một đường dẫn cụ thể.
func openStore(path string) (*Store, error) {
	store := &Store{path: path}
	if err := store.load(); err != nil {
		return nil, err
	}
	return store, nil
}

// load đọc danh sách đã lưu.
func (s *Store) load() error {
	data, err := os.ReadFile(s.path)
	if errors.Is(err, os.ErrNotExist) {
		s.list = []Server{}
		return nil
	}
	if err != nil {
		return err
	}

	if err := json.Unmarshal(data, &s.list); err != nil {
		// Tệp hỏng không được làm ứng dụng không mở lên được: bắt đầu lại với
		// danh sách rỗng còn hơn bắt người dùng đi sửa JSON bằng tay.
		s.list = []Server{}
	}

	// Bản ghi lưu từ phiên bản trước chỉ có panel nên không ghi trường kind.
	for i := range s.list {
		if s.list[i].Kind == "" {
			s.list[i].Kind = KindPanel
		}
	}
	return nil
}

// save ghi danh sách xuống đĩa.
//
// Quyền 0600 vì địa chỉ panel có kèm đường dẫn bí mật — thứ đứng giữa người lạ
// và trang đăng nhập.
func (s *Store) save() error {
	data, err := json.MarshalIndent(s.list, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(s.path, data, 0o600)
}

// List trả về danh sách máy chủ.
func (s *Store) List() []Server {
	s.mu.Lock()
	defer s.mu.Unlock()

	out := make([]Server, len(s.list))
	copy(out, s.list)
	return out
}

// Save thêm mới hoặc cập nhật một máy chủ.
func (s *Store) Save(server Server) (Server, error) {
	clean, err := normalizeServer(server)
	if err != nil {
		return Server{}, err
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	if clean.ID == "" {
		clean.ID = nextID(s.list)
		s.list = append(s.list, clean)
		return clean, s.save()
	}

	for i := range s.list {
		if s.list[i].ID != clean.ID {
			continue
		}
		// Vân tay khóa máy chủ chỉ còn đúng khi vẫn là máy đó; đổi địa chỉ hay
		// cổng nghĩa là đang trỏ đi nơi khác.
		if s.list[i].Host == clean.Host && s.list[i].sshPort() == clean.sshPort() {
			clean.Fingerprint = s.list[i].Fingerprint
		}
		clean.Last = s.list[i].Last
		s.list[i] = clean
		return clean, s.save()
	}
	return Server{}, errors.New("không tìm thấy máy chủ")
}

// normalizeServer kiểm và dọn một bản ghi trước khi lưu.
func normalizeServer(server Server) (Server, error) {
	server.Name = strings.TrimSpace(server.Name)

	if server.Kind == KindSSH {
		server.Host = strings.TrimSpace(server.Host)
		server.User = strings.TrimSpace(server.User)
		if server.Host == "" {
			return Server{}, errors.New("thiếu địa chỉ máy chủ")
		}
		if strings.ContainsAny(server.Host, " /\\") {
			return Server{}, errors.New("địa chỉ máy chủ không hợp lệ")
		}
		if server.User == "" {
			return Server{}, errors.New("thiếu tên đăng nhập")
		}
		if server.Port < 0 || server.Port > 65535 {
			return Server{}, errors.New("cổng phải nằm trong khoảng 1–65535")
		}
		if server.Port == 0 {
			server.Port = 22
		}
		server.KeyPath = strings.TrimSpace(server.KeyPath)
		server.URL = ""
		if server.Name == "" {
			server.Name = server.User + "@" + server.Host
		}
		return server, nil
	}

	address, err := normalizeURL(server.URL)
	if err != nil {
		return Server{}, err
	}
	server.Kind = KindPanel
	server.URL = address
	server.Host, server.Port, server.User = "", 0, ""
	server.KeyPath, server.Password, server.Fingerprint = "", "", ""
	if server.Name == "" {
		server.Name = address
	}
	return server, nil
}

// RememberFingerprint ghi nhận vân tay khóa máy chủ sau lần kết nối đầu tiên.
func (s *Store) RememberFingerprint(id, fingerprint string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	for i := range s.list {
		if s.list[i].ID == id {
			if s.list[i].Fingerprint == fingerprint {
				return nil
			}
			s.list[i].Fingerprint = fingerprint
			return s.save()
		}
	}
	return errors.New("không tìm thấy máy chủ")
}

// ByID tìm một máy chủ theo định danh.
func (s *Store) ByID(id string) (Server, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()

	for _, item := range s.list {
		if item.ID == id {
			return item, true
		}
	}
	return Server{}, false
}

// Remove xóa một máy chủ khỏi danh sách.
func (s *Store) Remove(id string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	out := s.list[:0]
	for _, item := range s.list {
		if item.ID != id {
			out = append(out, item)
		}
	}
	s.list = out
	return s.save()
}

// MarkLast ghi nhớ máy chủ vừa mở để lần chạy sau tự vào thẳng.
func (s *Store) MarkLast(id string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	for i := range s.list {
		s.list[i].Last = s.list[i].ID == id
	}
	return s.save()
}

// Last trả về máy chủ mở gần nhất.
func (s *Store) Last() (Server, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()

	for _, item := range s.list {
		if item.Last {
			return item, true
		}
	}
	return Server{}, false
}

// normalizeURL kiểm và chuẩn hóa địa chỉ panel.
func normalizeURL(raw string) (string, error) {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return "", errors.New("thiếu địa chỉ panel")
	}

	// Thiếu giao thức thì hiểu là http: panel mặc định chạy HTTP cho tới khi
	// người dùng gắn chứng chỉ, và bắt gõ đủ "http://" chỉ tạo thêm chỗ sai.
	if !strings.Contains(raw, "://") {
		raw = "http://" + raw
	}

	parsed, err := url.Parse(raw)
	if err != nil {
		return "", errors.New("địa chỉ không hợp lệ")
	}
	if parsed.Scheme != "http" && parsed.Scheme != "https" {
		return "", errors.New("địa chỉ phải bắt đầu bằng http:// hoặc https://")
	}
	if parsed.Host == "" {
		return "", errors.New("địa chỉ thiếu tên máy chủ")
	}

	// Đường dẫn bí mật phải kết thúc bằng dấu gạch chéo, nếu không thẻ base của
	// panel phân giải sai và giao diện nạp tài nguyên từ thư mục cha.
	if parsed.Path != "" && !strings.HasSuffix(parsed.Path, "/") {
		parsed.Path += "/"
	}
	parsed.Fragment = ""
	return parsed.String(), nil
}

// nextID sinh định danh chưa dùng.
func nextID(list []Server) string {
	next := 1
	for {
		candidate := "s" + strconv.Itoa(next)
		used := false
		for _, item := range list {
			if item.ID == candidate {
				used = true
				break
			}
		}
		if !used {
			return candidate
		}
		next++
	}
}
